src/k8s/pkg/k8sd/app/cluster_util.go

```diff
@@ -0,0 +1,131 @@
+package app
+
+import (
+	"fmt"
+	"net"
+	"path"
+
+	"github.com/canonical/k8s/pkg/k8sd/api/impl"
+	"github.com/canonical/k8s/pkg/k8sd/pki"
+	"github.com/canonical/k8s/pkg/k8sd/setup"
+	"github.com/canonical/k8s/pkg/k8sd/types"
+	"github.com/canonical/k8s/pkg/snap"
+	snaputil "github.com/canonical/k8s/pkg/snap/util"
+	"github.com/canonical/k8s/pkg/utils/k8s"
+	"github.com/canonical/microcluster/state"
+)
+
+func generateAndEnsureCertificates(snap snap.Snap, cfg types.ClusterConfig, nodeName string, allowSelfSignedCA bool) error {
```

Comment:
also update caller sites

```suggestion
func setupDatastoreCertificates(snap snap.Snap, cfg types.ClusterConfig, nodeName string, allowSelfSignedCA bool) error {
```